### PR TITLE
[tune] Fix an edge case where `DurableTrainable` would not delete checkpoints in remote storage

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

If both the runner and trial are present on the same node, `DurableTrainable` will never remove files in remote storage as it would be unable to find local files, which had already been deleted by the `checkpoint_deleter`. This change ensures that `checkpoint_deleter` doesn't remove local files by itself if the runner is on the same node.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by xwjiang2010** [comment]
How about instead, we change the logic in DurableTrainable to always remove checkpoint in remote storage whether or not local copy is found?
try:
except:
finally:

**Comment by Yard1** [comment]
@xwjiang2010 It won't be able to find the checkpoint directory through the method then

**Comment by xwjiang2010** [comment]
> @xwjiang2010 It won't be able to find the checkpoint directory through the find method then

Yeah that's another thing I don't quite understand. `find_checkpoint_dir` deals with some nested checkpoint case and recursively tries to find the top layer directory that contains `.is_checkpoint`. For the cases I have seen, it always gives me the same output as input tho. Would be good to understand the case this function is trying to address.

**Comment by xwjiang2010** [comment]
I am ok with the direction of the PR. Thanks Antoni for spotting the bug :)

Besides my question about when to care about nested checkpoint and thus `find_checkpoint_dir`, if we want to proceed with this direction, can we:
- Add some high level description of the responsibilities of CheckpointManager, DurableTrainable including in the context of when Trainable and CheckpointManager happen to be on the same physical machine. This would greatly help with reasoning about the code.
- Add a test?

File: python/ray/tune/trial.py [hunk 1]
```diff
@@ -80,23 +81,42 @@ def validate(formats):
...
+
+    @property
+    def runner_ip(self):
+        if not self._runner_ip:
+            try:
+                self._runner_ip = ray.get(
+                    self.runner.get_current_ip.remote(), timeout=30)
+            except GetTimeoutError:
+                pass
+        return self._runner_ip
```
Comment (krfricke) [inline python/ray/tune/trial.py:101]:
```suggestion
    def __init__(self, trial_id, runner, node_ip, timeout: int = 15):
        self.trial_id = trial_id
        self.runner = runner
        self.node_ip = node_ip
        self._runner_ip = None
        self.timeout = timeout

    @property
    def runner_ip(self):
        if not self._runner_ip:
            try:
                self._runner_ip = ray.get(
                    self.runner.get_current_ip.remote(), timeout=self.timeout)
            except GetTimeoutError:
                pass
        return self._runner_ip
```

**Review by krfricke — APPROVED** [review]
Thanks, I think we can go ahead with this. But I agree we might want to refactor this (remote) checkpointing logic later

File: python/ray/tune/trial.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 
 import ray
 import ray.cloudpickle as cloudpickle
+from ray.exceptions import GetTimeoutError
```
Comment (richardliaw) [inline python/ray/tune/trial.py:16]: Is there some way we can test this PR?

Comment (Yard1) [reply python/ray/tune/trial.py:16]: We could include a test for this in one of the release tests using s3, or we could use a library like moto to mock s3, though that may take some effort to set up

Comment (richardliaw) [reply python/ray/tune/trial.py:16]: OK. How about let's keep track of this in an issue, as we should add multi-testing soon.

Comment (xwjiang2010) [reply python/ray/tune/trial.py:16]: Do you know why the release test set up cannot access S3?

**Pull request merged**